katsdptelstate/telescope_state.py

```diff
@@ -744,6 +744,16 @@ def keys(self, filter='*'):
         keys = self._backend.keys(_ensure_binary(filter))
         return sorted(_ensure_str(key) for key in keys)
 
+    def _ipython_key_completions_(self):
+        """List of keys used in IPython (version >= 5) tab completion.
+
+        This respects the prefix list and presents keys with prefixes removed.
+        """
+        keys = []
+        for prefix in self._prefixes:
+            keys.extend(k[len(prefix):] for k in self.keys(prefix + b'*'))
+        return keys
+
```

Comment:
This could contain duplicates if a key appears in more than one namespace. Does it need to remove duplicates or does IPython handle that?

Reply:
That is the intended behaviour.

For me, tab completion is all about highlighting the contents of the container and it is more informative to list both the short and long names of a key. I've also ensured that the more specific namespaces are at the start of the list and the more generic ones at the end. Duplication is not an issue, since only a human will iterate over the options and make a choice.

I've even considered adding "spacer" completions between namespaces, something like `------------` or `---<namespace>---`, but this is complicated by `prompt_toolkit` listing the completions in a 2-D grid structure instead of readline's linear list.